Offline bootstrap for chocolatey

twangboy opened the pull request:
### What does this PR do?
Adds the ability to install Chocolatey by specifying a source.
Adds a dotnet salt util (`salt.utils.win_dotnet`) for working with versions. Includes the following functions:
- `dotnet.versions`
- `dotnet.versions_list`
- `dotnet.versions_details`
- `dotnet.version_atleast`

### What issues does this PR fix or reference?
https://github.com/saltstack/salt/issues/45969

### Tests written?
No

### Commits signed with GPG?
Yes

twangboy committed:
Add util to detect dotnet version

twangboy committed:
Add ability to install offline

Adds source parameter to bootstrap
Adds new function unbootstrap
Adds new function to win_dotnet version_atleast

twangboy committed:
Fix cmd.script

twangboy committed:
Sync running environment with system environment

twangboy committed:
Sync current environment on bootstrap

twangboy committed:
Check var exists before popping

twangboy committed:
Don't try to modify the parent environment

garethgreenaway commented:
@twangboy Looks like there is a lint error.

twangboy committed:
Fix lint

twangboy committed:
Add documentation to the unbootstrap function

twangboy committed:
Add docs for the win_dotnet salt util

twangboy committed:
Add unit tests

twangboy committed:
Fix some lint

twangboy committed:
Add versionadded

twangboy committed:
Fix docs, add release notes

cachedout commented on doc/topics/releases/neon.rst:
```diff
@@ -323,13 +323,31 @@ Module Changes
 - The :py:func:`yumpkg <salt.modules.yumpkg>` module has been updated to support
   VMWare's Photon OS, which uses tdnf (a C implementation of dnf).
 
+- The :py:func:`chocolatey.bootstrap <salt.modules.chocolatey.bootstrap>` function
```
Missing period at end of sentence.

cachedout commented on doc/topics/releases/neon.rst:
```diff
@@ -323,13 +323,31 @@ Module Changes
+  has been updated to support offline installation
+
+- The :py:func:`chocolatey.unbootstrap <salt.modules.chocolatey.unbootstrap>` function
+  has been added to uninstall Chocolatey
```
Also missing punctuation.

cachedout commented on doc/topics/releases/neon.rst:
```diff
@@ -323,13 +323,31 @@ Module Changes
+Util Changes
+============
+
+- The :py:func:`win_dotnet <salt.utils.win_dotnet>` salt util has been added to
```
s/Salt/salt/

cachedout commented on salt/modules/chocolatey.py:
```diff
@@ -109,6 +111,10 @@ def _find_chocolatey(context, salt):
     return choc_path
 
 
+def _dotnet_versions():
```
What's the purpose of this? I don't see it called anywhere.

cachedout commented on salt/modules/chocolatey.py:
```diff
@@ -140,18 +146,47 @@ def bootstrap(force=False):
+
+        source (str):
+            The location of the ``.nupkg`` file or ``.ps1`` file to run from an
+            alternate location. This can be one of the following types of urls:
```
URLs

cachedout commented on salt/modules/chocolatey.py:
```diff
@@ -161,71 +196,191 @@ def bootstrap(force=False):
+                      ''.format(__grains__['osrelease'])
+                if source:
+                    err = '{0}: PowerShell is required to bootstrap ' \
+                          'Chocolatey with Source'.format(err)
```
This would likely read more cleanly if you didn't reuse `err`.

cachedout commented on salt/utils/win_dotnet.py:
```diff
@@ -0,0 +1,157 @@
+    return versions()['details']
+
+
+def version_atleast(version):
```
This seems like it should be `version_at_least` to be consistent with breaking between words.

twangboy committed:
Add Mike's suggestions

Fix some punctuation
Change version_atleast to version_at_least
Remove unused hidden function _dotnet_versions
Don't reuse err

twangboy committed:
Fix a typo

twangboy committed:
Merge branch 'develop' into offline_bootstrap

twangboy committed:
Merge branch 'develop' into offline_bootstrap

twangboy committed:
Merge branch 'develop' into offline_bootstrap

twangboy committed:
Merge branch 'develop' into offline_bootstrap
